### PR TITLE
PMM-5839 Fixed incorrect metrics collection.

### PR DESCRIPTION
**Pull request by BupycHuk**

### CODE REVIEW

**Comment by BupycHuk** [comment]
@AlekSi It's not possible to write separate test with a creating mock table with data, selecting data and checking m_blk_read_time. Postgres keeps data in memory and doesn't read from the disk, that's why block read time is equal to zero.

**Comment by AlekSi** [comment]
@BupycHuk CI fails

File: agents/postgres/pgstatstatements/pgstatstatements.go [hunk 1]
```diff
@@ -242,8 +242,8 @@ func makeBuckets(current, prev map[int64]*pgStatStatementsExtended, l *logrus.En
 			{float32(currentPSS.TempBlksRead - prevPSS.TempBlksRead), &mb.Postgresql.MTempBlksReadSum, &mb.Postgresql.MTempBlksReadCnt},
 			{float32(currentPSS.TempBlksWritten - prevPSS.TempBlksWritten), &mb.Postgresql.MTempBlksWrittenSum, &mb.Postgresql.MTempBlksWrittenCnt},
 
-			{float32(currentPSS.BlkReadTime - prevPSS.BlkReadTime), &mb.Postgresql.MBlkReadTimeSum, &mb.Postgresql.MBlkReadTimeCnt},
-			{float32(currentPSS.BlkWriteTime - prevPSS.BlkWriteTime), &mb.Postgresql.MBlkWriteTimeSum, &mb.Postgresql.MBlkWriteTimeCnt},
+			{float32(currentPSS.BlkReadTime-prevPSS.BlkReadTime) / 1000, &mb.Postgresql.MBlkReadTimeSum, &mb.Postgresql.MBlkReadTimeCnt},
```
Comment (AlekSi) [inline agents/postgres/pgstatstatements/pgstatstatements.go:245]: Let's add a comment as above:
```
// convert milliseconds to seconds
```

**Comment by AlekSi** [comment]
@BupycHuk I think we should tweak tests, not just restart them

**Comment by BupycHuk** [comment]
I'm trying to figure out is this because of changes or it's just unstable. Locally it works fine.

**Comment by AlekSi** [comment]
This is similar to how our QAN data collection tests failed initially – we want them to be precise as much as possible, but they can't be fully precise by nature. We don't want to completely skip size checks and don't want to use a too wide range, but we have to tweak every now and then for some time.

File: actions/postgresql_query_show_action_test.go [hunk 2]
```diff
@@ -45,13 +45,13 @@ func TestPostgreSQLQueryShow(t *testing.T) {
 
 		b, err := a.Run(ctx)
 		require.NoError(t, err)
-		assert.LessOrEqual(t, 22151, len(b))
+		assert.GreaterOrEqual(t, len(b), 22150)
 		assert.LessOrEqual(t, len(b), 29156)
```
Comment (AlekSi) [inline actions/postgresql_query_show_action_test.go:49]: It was done intentionally: `22151 < len(b) < 29156`

**Comment by AlekSi** [comment]
Does discussion in Slack change anything?

File: actions/postgresql_query_show_action_test.go (hunk 2 above)
Comment (BupycHuk) [reply actions/postgresql_query_show_action_test.go:49]: And it makes it more clear to understand that len(b) should be greater then 22150 and less then 29156

**Comment by BupycHuk** [comment]
@AlekSi yes, I figured out how to make a test

File: actions/postgresql_query_show_action_test.go (hunk 2 above)
Comment (AlekSi) [reply actions/postgresql_query_show_action_test.go:49]: Well, I don't agree that it makes it more clear: I read origin code as `22151 <= len(b) <= 29156` (as one would write in Python or just in math) and new code as `len(b) >= 22150 && len(b) <= 29156`.

But if you insist, please change that in other files and tests.

File: agents/postgres/pgstatstatements/pgstatstatements_test.go [hunk 3]
```diff
@@ -295,5 +309,86 @@ func TestPGStatStatementsQAN(t *testing.T) {
...
+	t.Run("CheckMBlkReadTime", func(t *testing.T) {
+		r := rand.New(rand.NewSource(time.Now().Unix()))
+		tableName := fmt.Sprintf("customer%d", r.Int())
+		_, err := db.Exec(fmt.Sprintf(`
+		CREATE TABLE %s (
+			customer_id integer NOT NULL,
+			first_name character varying(45) NOT NULL,
+			last_name character varying(45) NOT NULL,
+			active boolean
+		);`, tableName))
```
Comment (AlekSi) [inline agents/postgres/pgstatstatements/pgstatstatements_test.go:324]: There is no real need to add `;` to queries

File: agents/postgres/pgstatstatements/pgstatstatements_test.go [hunk 4]
```diff
@@ -295,5 +309,86 @@ func TestPGStatStatementsQAN(t *testing.T) {
...
+		require.NoError(t, err)
+		defer func() {
+			_, err := db.Exec(fmt.Sprintf(`DROP TABLE %s;`, tableName))
+			require.NoError(t, err)
+		}()
+		m := setup(t, db)
+
+		waitGroup := sync.WaitGroup{}
```
Comment (AlekSi) [inline agents/postgres/pgstatstatements/pgstatstatements_test.go:332]: That's something I say over and over again: there is no field you can possible initialise in WaitGroup, so please use `var wg sync.WaitGroup`.

And I think we should add go-consistent to this repo.

File: actions/postgresql_query_show_action_test.go (hunk 2 above)
Comment (BupycHuk) [reply actions/postgresql_query_show_action_test.go:49]: Ok, reverted it.

File: agents/postgres/pgstatstatements/pgstatstatements_test.go (hunk 4 above)
Comment (BupycHuk) [reply agents/postgres/pgstatstatements/pgstatstatements_test.go]: I'll create task for go-consistent

File: agents/postgres/pgstatstatements/pgstatstatements_test.go [hunk 5]
```diff
@@ -295,5 +309,86 @@ func TestPGStatStatementsQAN(t *testing.T) {
...
+		_, err := db.Exec(fmt.Sprintf(`
+		CREATE TABLE %s (
+			customer_id integer NOT NULL,
+			first_name character varying(45) NOT NULL,
+			last_name character varying(45) NOT NULL,
+			active boolean
+		)`, tableName))
+		require.NoError(t, err)
+		defer func() {
+			_, err := db.Exec(fmt.Sprintf(`DROP TABLE %s;`, tableName))
```
Comment (AlekSi) [inline agents/postgres/pgstatstatements/pgstatstatements_test.go:327]: `;`

**Pull request merged**